zerver/actions/presence.py

```diff
@@ -77,39 +81,73 @@ def do_update_user_presence(
 ) -> None:
     client = consolidate_client(client)
 
+    # TODO: While we probably DO want creating an account to
+    # automatically create a first `UserPresence` object with
+    # last_connected_time and last_active_time as the current time,
+    # our presence tests don't understand this, and it'd be perhaps
+    # wrong for some cases of account creation via the API.  So we may
+    # want a "never" value here as the default.
     defaults = dict(
-        timestamp=log_time,
-        status=status,
+        # Given that these are defaults for creation of a UserPresence row
+        # if one doesn't yet exist, the most sensible way to do this
+        # is to set both last_active_time and last_connected_time
+        # to log_time.
+        last_active_time=log_time,
+        last_connected_time=log_time,
         realm_id=user_profile.realm_id,
     )
+    if status == UserPresence.LEGACY_STATUS_IDLE_INT:
+        # If the presence entry for the user is just to be created, and
+        # we want it to be created as idle, then there needs to be an appropriate
+        # offset between last_active_time and last_connected_time, since that's
+        # what the event-sending system calculates the status based on, via
+        # format_legacy_presence_dict.
+        defaults["last_active_time"] = log_time - datetime.timedelta(
+            seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS + 1
+        )
 
     (presence, created) = UserPresence.objects.get_or_create(
         user_profile=user_profile,
-        client=client,
         defaults=defaults,
     )
 
-    stale_status = (log_time - presence.timestamp) > datetime.timedelta(minutes=1, seconds=10)
-    was_idle = presence.status == UserPresence.IDLE
-    became_online = (status == UserPresence.ACTIVE) and (stale_status or was_idle)
+    time_since_last_active = log_time - presence.last_active_time
+    now_online = time_since_last_active > datetime.timedelta(
+        # With the default tuning of these constants, where OFFLINE_THRESHOLD_SECS
+        # is equal to PRESENCE_PING_INTERVAL_SECS * 3 + 20, the expression below
+        # is equivalent to just 2 * PRESENCE_PING_INTERVAL_SECS + 10, so just enough
+        # time for the client to have missed two pings plus a small extra margin.
+        seconds=settings.OFFLINE_THRESHOLD_SECS
+        - settings.PRESENCE_PING_INTERVAL_SECS
+        - 10
+    )
```

Comment:
Tweaked this to address https://github.com/zulip/zulip/pull/16381/commits/135d46f2c78c1679bb46ccbab0f58df295073ca1#r1101923300 and attempted to add a comment giving some clarification, so that this doesn't become a mysterious magical arithmetic expression in the future - but not sure what I wrote there is enough - feel free to perhaps add more justification if appropriate

Reply:
Not having followed this conversation, this comment leaves me wondering "so why not write `2 * PRESENCE_PING_INTERVAL_SECS + 10`?"

Reply:
I think the reason we're using this formula is that we're sending this notice if the client is at risk of demoting the user to offline before the next time it would query the server via a regular checkin. Probably the comment should reflect that.

Reply:
Maybe:

> Here, we decide whether the user is newly online, and we should consider sending an immediate presence update via the events system that this user is now online, rather than waiting for other clients to poll the presence update. Immediate presence updates add load to the system, so we only want to do this if the user has missed a couple regular presence checkins (so their state is at least `2 * PRESENCE_PING_INTERVAL_SECS + 10` old), and also is risk of being shown by clients as offline before their next regular presence checkin (so at least `settings.OFFLINE_THRESHOLD_SECS - settings.PRESENCE_PING_INTERVAL_SECS - 10`). These two values happen to be the same in the default configuration.

Does it make sense to write this block using `max(2 * PRESENCE_PING_INTERVAL_SECS + 10, settings.OFFLINE_THRESHOLD_SECS - settings.PRESENCE_PING_INTERVAL_SECS - 10)`? 

Or maybe better we should write `assert OFFLINE_THRESHOLD_SECS >= 3 * settings.PRESENCE_PING_INTERVAL_SECS + 20` here next to a comment of this form and leave it roughly as written; I think it would not be a happy state if that were not true given this behavior.

Reply:
Added the comment with tiny phrasing tweaks, and went with adding the assert. Kinda ugly with the exclusion for tests:
```
    if not settings.TEST_SUITE:  # nocoverage
        # We exclude the test suite from the assertions, since in tests we may want to
        # mock settings arbitrarily.
        assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
```

the exclusion is not really necessary since it's only one existing test that breaks the assumption so it could be fixed - but I feel this might be the right choice in general so that any future tests can be written mocking these values freely without crashing into an annoying assertion?

Reply:
I don't like `not TEST_SUITE` assertions since they make the test suite not predictive of what will happen in production. In particular, if we letter changes these values and don't remember this line of code, we could merge code that fails an assertion every time it runs :)

It would be better to change the test, and have a comment here the explains what we're doing with the math.

Reply:
Cool, the test didn't really need to  any settings, so just squashed this tweak into the original commit:
```diff
diff --git a/zerver/actions/presence.py b/zerver/actions/presence.py
index cf480c6076..392639b9f7 100644
--- a/zerver/actions/presence.py
+++ b/zerver/actions/presence.py
@@ -120,10 +120,7 @@ def do_update_user_presence(
         # Same approach as above.
         time_since_last_connected = datetime.timedelta(days=1)
 
-    if not settings.TEST_SUITE:  # nocoverage
-        # We exclude the test suite from the assertions, since in tests we may want to
-        # mock settings arbitrarily.
-        assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
+    assert (3 * settings.PRESENCE_PING_INTERVAL_SECS + 20) <= settings.OFFLINE_THRESHOLD_SECS
     now_online = time_since_last_active > datetime.timedelta(
         # Here, we decide whether the user is newly online, and we need to consider
         # sending an immediate presence update via the events system that this user is now online,
diff --git a/zerver/tests/test_presence.py b/zerver/tests/test_presence.py
index b2f0a0c72e..d3767d4a83 100644
--- a/zerver/tests/test_presence.py
+++ b/zerver/tests/test_presence.py
@@ -630,8 +630,17 @@ class UserPresenceAggregationTests(ZulipTestCase):
         user = self.example_user("othello")
         self.login_user(user)
         validate_time = timezone_now()
-        with self.settings(OFFLINE_THRESHOLD_SECS=1):
-            result_dict = self._send_presence_for_aggregated_tests(user, "idle", validate_time)
+        result_dict = self._send_presence_for_aggregated_tests(user, "idle", validate_time)
+
+        with mock.patch(
+            "zerver.views.presence.timezone_now",
+            return_value=validate_time + datetime.timedelta(settings.OFFLINE_THRESHOLD_SECS + 1),
+        ):
+            # After settings.OFFLINE_THRESHOLD_SECS + 1 this generated, recent presence data
+            # will count as offline.
+            result = self.client_get(f"/json/users/{user.email}/presence")
+        result_dict = self.assert_json_success(result)
+
         self.assertDictEqual(
             result_dict["presence"]["aggregated"],
             {

```